annotatedcsv/decode_test.go

```diff
@@ -0,0 +1,844 @@
+		Name  string    `flux:"Name"`
+		Score float64   `flux:"Score"`
+		Sum   float64
+	}{
+		Index: 0,
+		Time:  time.Time{},
+		Score: 0.0,
+		Name:  "",
+	}, sn)
+}
+
+func TestDecodeFail(t *testing.T) {
+	csvTable := `#datatype,long,double,dateTime:RFC3339Nano,stringer
+#default,,,,
+,index,score,time,name
+,0,3.3,2021-02-18T10:34:08.135814545Z,Thomas
+,1,5.1,2021-02-18T22:08:44.850214724Z,John
+	
+`
+	reader := strings.NewReader(csvTable)
+	res := NewReader(reader)
+	require.True(t, res.NextSection())
+	require.NoError(t, res.Err())
+	require.True(t, res.NextRow())
+	require.NoError(t, res.Err())
+
+	err := res.Decode(map[string]int{})
+	require.Error(t, err)
+
+	s := struct {
+		Index int64     `flux:"Index"`
+		Time  time.Time `flux:"Time"`
+		Name  string    `flux:"Name"`
+		Score float64   `flux:"Score"`
+		Sum   float64
+	}{}
+
+	err = res.Decode(s)
+	require.Error(t, err)
+
+	var r []float64
+	err = res.Decode(r)
+	require.Error(t, err)
+
+	err = res.Decode(&r)
+	require.Error(t, err)
+
+	var f float64
+	err = res.Decode(&f)
+	require.Error(t, err)
+
+}
+
+func TestDecodeSliceBasicAllTypes(t *testing.T) {
+	csvTable := `#datatype,string,unsignedLong,dateTime:RFC3339,dateTime:RFC3339,dateTime:RFC3339Nano,duration,string,long,base64Binary,boolean
+#group,false,false,true,true,false,false,true,true,true,true
+#default,_result,,,,,,,,,
+,result,table,_start,_stop,_time,took,_field,index,note,b
+,,0,2020-02-17T22:19:49.747562847Z,2020-02-18T22:19:49.747562847Z,2020-02-18T10:34:08.135814545Z,32m,f,-1,ZGF0YWluYmFzZTY0,true
+,,0,2020-02-17T22:19:49.747562847Z,2020-02-18T22:19:49.747562847Z,2020-02-18T22:08:44.850214724Z,1h23m4s,f,1,eHh4eHhjY2NjY2NkZGRkZA==,false
+
+#datatype,long,double,dateTime,string
+#default,,,,
+,index,score,time,name
+,0,3.3,2021-02-18T10:34:08.135814545Z,Thomas
+,1,5.1,2021-02-18T22:08:44.850214724Z,John
+
+`
+	reader := strings.NewReader(csvTable)
+	res := NewReader(reader)
+	require.True(t, res.NextSection())
+	require.NoError(t, res.Err())
+	require.True(t, res.NextRow())
+	require.NoError(t, res.Err())
+
+	var r []interface{}
+	require.NoError(t, res.Decode(&r))
+	er := []interface{}{
+		"_result",
+		uint64(0),
+		mustParseTime("2020-02-17T22:19:49.747562847Z"),
+		mustParseTime("2020-02-18T22:19:49.747562847Z"),
+		mustParseTime("2020-02-18T10:34:08.135814545Z"),
+		time.Minute * 32,
+		"f",
+		int64(-1),
+		[]byte("datainbase64"),
+		true,
+	}
+	require.Equal(t, er, r)
+
+	var rs []string
+	require.NoError(t, res.Decode(&rs))
+	er2 := []string{
+		"_result",
+		"0",
+		"2020-02-17T22:19:49.747562847Z",
+		"2020-02-18T22:19:49.747562847Z",
+		"2020-02-18T10:34:08.135814545Z",
+		"32m",
+		"f",
+		"-1",
+		"ZGF0YWluYmFzZTY0",
+		"true",
+	}
+	require.Equal(t, er2, rs)
+}
+
+func TestConversionErrors(t *testing.T) {
+	s1 := &struct {
+		S []byte
+		D time.Duration
+		T time.Time
+		I int8
+		U uint8
+		B bool
+		F float32
+	}{}
+	require.Error(t, toBytes(reflect.ValueOf(s1).Elem().FieldByName("S"), "#"))
```

Comment:
I'd much prefer it if all these tests were done through the public API, please.

That way the tests give us full confidence that the overall logic works, and we have freedom to refactor internally without needing to change the tests.

I'd also like to see the actual error string in the tests. Too many times I've found tests that are testing for just the existence of an error but the tests were passing with an actual error that is either unreadable or misleading or just not the error that's expected.

A lot of this looks like it would fit a table-driven test very well:

```
var singleRowDecodeTests = struct{
    testName string
    csv string
    value interface{}
    expectError string
}{{
   testName: "Simple",
   csv: `#datatype,long,double,dateTime,string
#default,,,,
,index,score,time,name
,0,3.3,2021-02-18T10:34:08.135814545Z,Thomas
`,
    value: struct{
         Score int `flux:"score"`
         Name string `flux:"name"`
    }{
        Score: 3,
        Name: "Thomas",
    },
}, {
    testName: "InvalidDuration",
    csv: `#datatype,duration
,d
,#
`,
     value: struct{
         D time.Duration `flux:"d"`
     },
     expectError: "cannot convert column d of type duration to time.Duration: invalid etc",
}}

func TestSingleRowDecode(t *testing.T) {
    for _, test := range singleRowDecodeTests {
         t.Run(test.testName, func(t *testing.T) {
             decVal := reflect.New(reflect.TypeOf(test.value))
             r := NewReader(strings.NewReader(test.csv))
             err := r.Decode(decVal.Interface())
             if test.expectError {
                 require.Error(t, err)
                 require.Equal(t, test.expectError, err.Error()
                 return
              }
              require.NoError(t, err)
              require.Equal(t, test.value, decVal.Elem().Interface())
         })
    }
}
```

You can probably cover a great deal of the functionality by populating that table, leaving other cases like reading multiple rows and different decode types for other more specific tests.

To reduce verbosity, a couple of helper functions might turn out useful (e.g. a function that returns CSV data with a single row with a given field type, field name and field value; a function that returns a struct containing a single field with a given type, value and flux tag).

Reply:
In fact, this is the only unit test testing by private API. It is done this way to easily cover all type combinations, even those not possible using supported column types (mostly negative tests).
As others are testing using public API by inputting CSV data, I thought, this test could just use private API to be concise as possible

Reply:
I'd still suggest doing this via the public API. It would be great to have as much freedom to refactor as possible. I think you should be able to get it just about as terse by using a bit of reflection.